crates/sui-graphql-rpc/src/types/event.rs

```diff
@@ -137,62 +139,98 @@ impl Event {
     ) -> Result<Connection<String, Event>, Error> {
-        let (prev, next, results) = db
-            .execute(move |conn| {
-                // TODO (wlmyng) the None value will be replaced by self.checkpoint_viewed_at
-                page.paginate_query::<StoredEvent, _, _, _>(conn, None, move || {
-                    let mut query = events::dsl::events.into_boxed();
-
-                    // The transactions table doesn't have an index on the senders column, so use
-                    // `tx_senders`.
-                    if let Some(sender) = &filter.sender {
-                        query = query.filter(
-                            events::dsl::tx_sequence_number.eq_any(
-                                tx_senders::dsl::tx_senders
-                                    .select(tx_senders::dsl::tx_sequence_number)
-                                    .filter(tx_senders::dsl::sender.eq(sender.into_vec())),
-                            ),
-                        )
-                    }
-
-                    if let Some(digest) = &filter.transaction_digest {
+        let cursor_viewed_at = validate_cursor_consistency(page.after(), page.before())?;
+        let checkpoint_viewed_at: Option<u64> = cursor_viewed_at.or(checkpoint_viewed_at);
+
+        let ((prev, next, results), checkpoint_viewed_at) = db
+            .execute_repeatable(move |conn| {
+                let checkpoint_viewed_at = match checkpoint_viewed_at {
+                    Some(value) => Ok(value),
+                    None => Checkpoint::available_range(conn).map(|(_, rhs)| rhs),
+                }?;
+
+                let result = page.paginate_query::<StoredEvent, _, _, _>(
+                    conn,
+                    Some(checkpoint_viewed_at),
+                    move || {
+                        let mut query = events::dsl::events.into_boxed();
+
+                        // Bound events by the provided `checkpoint_viewed_at`.
                         query = query.filter(
-                            events::dsl::tx_sequence_number.eq_any(
-                                transactions::dsl::transactions
-                                    .select(transactions::dsl::tx_sequence_number)
-                                    .filter(
-                                        transactions::dsl::transaction_digest.eq(digest.to_vec()),
-                                    ),
-                            ),
-                        )
-                    }
-
-                    if let Some(module) = &filter.emitting_module {
-                        query = module.apply(query, events::dsl::package, events::dsl::module);
-                    }
-
-                    if let Some(type_) = &filter.event_type {
-                        query = type_.apply(query, events::dsl::event_type);
-                    }
-
-                    query
-                })
+                            events::dsl::checkpoint_sequence_number.le(checkpoint_viewed_at as i64),
```

Comment:
This seems like it would have similar problems to the checkpoint-based queries on transaction -- can you check that this doesn't hurt performance, and if it does, switch to a transaction sequence number based bound?

Reply:
Ah yes good call here. indeed it makes the query perform much worse. We can translate the checkpoint into
```
select tx_sequence_number, event_sequence_number from events where tx_sequence_number <= (select tx_sequence_number from transactions where checkpoint_sequence_number < 3000000 order by tx_sequence_number desc limit 1) order by tx_sequence_number desc, event_sequence_number desc limit 50;
```